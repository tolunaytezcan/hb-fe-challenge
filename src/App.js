import Header from 'components/organism/Header/Header';
import SearchResult from 'components/organism/SearchResult/SearchResult';

import 'styles/globals.scss';

const App = () => (
  <>
    <Header />
    <SearchResult />
  </>
);

export default App;
