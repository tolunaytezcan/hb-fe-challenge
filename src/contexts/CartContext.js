import { createContext, useState } from 'react';

const CartContext = createContext();

export default CartContext;

export const CartProvider = ({ children }) => {
  const [cartProducts, setCartProducts] = useState([]);

  const addProductToCart = newProduct => {
    if (!cartProducts.some(item => item.id === newProduct.id))
      return setCartProducts([...cartProducts, newProduct]);
  };

  const removeProductFromCart = productId => {
    return setCartProducts(cartProducts.filter(item => item.id !== productId));
  };

  const values = {
    cartProducts,
    setCartProducts,
    addProductToCart,
    removeProductFromCart
  };

  return <CartContext.Provider value={values}>{children}</CartContext.Provider>;
};
