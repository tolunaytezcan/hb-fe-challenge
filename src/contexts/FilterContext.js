import { createContext, useState, useEffect, useMemo } from 'react';
import { useHistory, useLocation } from 'react-router-dom';

const FilterContext = createContext();

export default FilterContext;

function useQuery() {
  const { search } = useLocation();
  return useMemo(() => new URLSearchParams(search), [search]);
}

export const FilterProvider = ({ children }) => {
  const history = useHistory();
  const [selectedFilters, setSelectedFilters] = useState({
    color: '',
    brand: '',
    sort: '',
    search: '',
    page: 1
  });

  const clearFilters = () => {
    setSelectedFilters({
      color: '',
      sort: '',
      search: '',
      brand: '',
      page: 1
    });
  };

  let query = useQuery();

  const colorQuery = query.get('color');
  const brandQuery = query.get('brand');
  const sortQuery = query.get('sort');
  const searchQuery = query.get('search');
  const pageQuery = query.get('page');

  useEffect(() => {
    setSelectedFilters(prev => ({ ...prev, brand: brandQuery }));
    setSelectedFilters(prev => ({ ...prev, color: colorQuery }));
    setSelectedFilters(prev => ({ ...prev, sort: sortQuery }));
    setSelectedFilters(prev => ({ ...prev, search: searchQuery }));
    setSelectedFilters(prev => ({ ...prev, page: +pageQuery }));
  }, []);

  useEffect(() => {
    history.push(
      `/?brand=${selectedFilters.brand}&color=${selectedFilters.color}&sort=${selectedFilters.sort}&search=${selectedFilters.search}&page=${selectedFilters.page}`
    );
  }, [selectedFilters]);

  const values = { selectedFilters, setSelectedFilters, clearFilters };

  return <FilterContext.Provider value={values}>{children}</FilterContext.Provider>;
};
