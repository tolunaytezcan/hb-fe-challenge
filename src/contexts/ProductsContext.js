import { useState, useEffect, createContext } from 'react';
import axios from 'axios';
import { baseURL } from 'api';

const ProductsContext = createContext();

export default ProductsContext;

export const ProductsProvider = ({ children }) => {
  const [products, setProducts] = useState([]);
  const [isLoading, setIsLoading] = useState(false);
  const [isError, setIsError] = useState(false);

  useEffect(() => {
    const getProducts = async () => {
      setIsLoading(true);
      try {
        const response = await axios.get(baseURL);
        setProducts(response.data);
      } catch (error) {
        setIsError(true);
      }
      setIsLoading(false);
    };

    !products.length > 0 && getProducts();
  }, [products]);

  const values = { products, isLoading, isError };

  return <ProductsContext.Provider value={values}>{children}</ProductsContext.Provider>;
};
