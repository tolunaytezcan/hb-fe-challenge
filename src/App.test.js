import { render } from '@testing-library/react';
import { BrowserRouter as Router } from 'react-router-dom';
import CartContext from 'contexts/CartContext';
import FilterContext from 'contexts/FilterContext';
import ProductsContext from 'contexts/ProductsContext';

import App from './App';

describe('Project should be worked without crash', () => {
  const products = [
    {
      id: 1,
      brand: 'Apple',
      title: 'iPhone 11 64GB',
      price: 13950,
      color: 'Siyah',
      images: [
        {
          image: 'https://productimages.hepsiburada.net/s/49/400-592/10986385899570.jpg'
        }
      ],
      discountRate: 12,
      createdDate: '2022-07-02T17:51:41.382Z'
    }
  ];

  const cartProducts = [
    {
      id: 1,
      brand: 'Apple',
      title: 'iPhone 12 128GB',
      price: 13950,
      color: 'Siyah',
      images: [
        {
          image: 'https://productimages.hepsiburada.net/s/49/400-592/10986385899570.jpg'
        }
      ],
      discountRate: 12,
      createdDate: '2022-07-02T17:51:41.382Z'
    }
  ];

  let selectedFilters = {
    color: '',
    brand: '',
    sort: '',
    search: '',
    page: 1
  };

  const setSelectedFilters = () => {
    selectedFilters = {
      color: '',
      brand: '',
      sort: '',
      search: '',
      page: 1
    };
  };

  it('App component', () => {
    const component = render(
      <Router>
        <FilterContext.Provider
          value={{
            selectedFilters,
            setSelectedFilters
          }}>
          <ProductsContext.Provider value={{ products }}>
            <CartContext.Provider value={{ cartProducts }}>
              <App />
            </CartContext.Provider>
          </ProductsContext.Provider>
        </FilterContext.Provider>
      </Router>
    );

    expect(component).toMatchSnapshot();
  });
});
