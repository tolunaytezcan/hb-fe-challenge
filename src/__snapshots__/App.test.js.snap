// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Project should be worked without crash should be rendered app component 1`] = `
Object {
  "asFragment": [Function],
  "baseElement": <body>
    <div>
      <div
        class="headerWrapper"
      >
        <div
          class="logoContainer"
        >
          <img
            alt="Logo icon"
            class="hbLogo"
            src="Logo.svg"
          />
        </div>
        <div
          class="searchInputContainer"
        >
          <img
            alt="Search icon"
            class="searchIcon"
            src="Search.svg"
          />
          <input
            class="searchInput"
            placeholder="25 milyon'dan fazla ürün içerisinde ara"
            type="text"
            value=""
          />
        </div>
        <div
          class="cartContainer"
        >
          <button
            class="buttonContainer  undefined"
          >
            <span>
              Sepetim
            </span>
            <span
              class="quantityBadge"
            >
              1
            </span>
          </button>
          <div
            class="cart"
          >
            <div
              class="cartContainer"
            >
              <div
                class="itemContainer"
              >
                <div
                  class="imageContainer"
                >
                  <img
                    alt="cart-product-image"
                    class="image"
                    src="https://productimages.hepsiburada.net/s/49/400-592/10986385899570.jpg"
                  />
                </div>
                <div
                  class="titleAndButtonContainer"
                >
                  <p
                    class="title"
                  >
                    iPhone 12 128GB
                  </p>
                  <button
                    class="buttonContainer chartRemove undefined"
                  >
                    <span>
                      Kaldır
                    </span>
                  </button>
                </div>
              </div>
            </div>
          </div>
        </div>
      </div>
      <div
        class="searchResultWrapper"
      >
        <div
          class="SubHeaderContainer"
        >
          <div
            class="SubHeader"
          >
            <span
              class="PageTitle"
            >
              Cep Telefonu Fiyatları ve Modelleri
            </span>
            <span
              class="SearchTextContainer"
            />
          </div>
          <div>
            <div
              class="SelectboxContainer"
            >
              <button
                class="buttonContainer  buttonIcon"
              >
                <span>
                  Sıralama
                </span>
                <img
                  alt="buttonIcon"
                  src="DownArrow.svg"
                />
              </button>
              <div
                class="SelectboxOptions"
              />
            </div>
          </div>
        </div>
        <div
          class="filterAndProducts"
        >
          <div
            class="sideBarContainer"
          >
            <div
              class="SidebarContainer"
            >
              <p
                class="FilterName"
              >
                Renk
              </p>
              <ul
                class="FilterList"
              >
                <li
                  class="Element "
                >
                  <span>
                    Siyah
                  </span>
                  <span>
                    (
                    1
                    )
                  </span>
                </li>
              </ul>
            </div>
            <div
              class="SidebarContainer"
            >
              <p
                class="SortName"
              >
                Sıralama
              </p>
              <ul
                class="SortList"
              />
            </div>
            <div
              class="SidebarContainer"
            >
              <p
                class="FilterName"
              >
                Marka
              </p>
              <ul
                class="FilterList"
              >
                <li
                  class="Element "
                >
                  <span>
                    Apple
                  </span>
                  <span>
                    (
                    1
                    )
                  </span>
                </li>
              </ul>
            </div>
          </div>
          <div>
            <div
              class="productListContainer"
            >
              <div
                class="productCardContainer"
              >
                <div
                  class="imageContainer"
                >
                  <img
                    class="image"
                    src="https://productimages.hepsiburada.net/s/49/400-592/10986385899570.jpg"
                  />
                </div>
                <div
                  class="productProperties"
                >
                  <div>
                    <p
                      class="productTitle"
                    >
                      iPhone 11 64GB
                    </p>
                    <p
                      class="productbrand"
                    >
                      <span>
                        Marka: 
                      </span>
                      Apple
                    </p>
                    <p
                      class="productColor"
                    >
                      <span>
                        Renk: 
                      </span>
                      Siyah
                    </p>
                    <p
                      class="discountedPrice"
                    >
                      12.276,00
                       
                      TL
                    </p>
                  </div>
                  <div
                    class="prevPriceContainer"
                  >
                    <span
                      class="prevPrice"
                    >
                      13.950,00
                       TL
                    </span>
                    <span
                      class="discountRate"
                    >
                      12
                      %
                    </span>
                  </div>
                  <div
                    class="productCartButton"
                  >
                    <button
                      class="buttonContainer orange undefined"
                      disabled=""
                    >
                      <span>
                        Bu ürün zaten sepetinizde ekli
                      </span>
                    </button>
                  </div>
                </div>
              </div>
            </div>
            <div>
              <div
                class="PaginationContainer"
              >
                <span
                  class="arrowButton FirstArrow"
                >
                  &lt;&lt;
                </span>
                <span
                  class="arrowButton"
                >
                  &lt;
                </span>
                <ul
                  class="PaginationList"
                >
                  <li
                    class="activeNumber"
                  >
                    <span
                      class="numberButton"
                    >
                      1
                    </span>
                  </li>
                </ul>
                <span
                  class="arrowButton"
                >
                  &gt;
                </span>
                <span
                  class="arrowButton LastArrow"
                >
                  &gt;&gt;
                </span>
              </div>
            </div>
          </div>
        </div>
      </div>
    </div>
  </body>,
  "container": <div>
    <div
      class="headerWrapper"
    >
      <div
        class="logoContainer"
      >
        <img
          alt="Logo icon"
          class="hbLogo"
          src="Logo.svg"
        />
      </div>
      <div
        class="searchInputContainer"
      >
        <img
          alt="Search icon"
          class="searchIcon"
          src="Search.svg"
        />
        <input
          class="searchInput"
          placeholder="25 milyon'dan fazla ürün içerisinde ara"
          type="text"
          value=""
        />
      </div>
      <div
        class="cartContainer"
      >
        <button
          class="buttonContainer  undefined"
        >
          <span>
            Sepetim
          </span>
          <span
            class="quantityBadge"
          >
            1
          </span>
        </button>
        <div
          class="cart"
        >
          <div
            class="cartContainer"
          >
            <div
              class="itemContainer"
            >
              <div
                class="imageContainer"
              >
                <img
                  alt="cart-product-image"
                  class="image"
                  src="https://productimages.hepsiburada.net/s/49/400-592/10986385899570.jpg"
                />
              </div>
              <div
                class="titleAndButtonContainer"
              >
                <p
                  class="title"
                >
                  iPhone 12 128GB
                </p>
                <button
                  class="buttonContainer chartRemove undefined"
                >
                  <span>
                    Kaldır
                  </span>
                </button>
              </div>
            </div>
          </div>
        </div>
      </div>
    </div>
    <div
      class="searchResultWrapper"
    >
      <div
        class="SubHeaderContainer"
      >
        <div
          class="SubHeader"
        >
          <span
            class="PageTitle"
          >
            Cep Telefonu Fiyatları ve Modelleri
          </span>
          <span
            class="SearchTextContainer"
          />
        </div>
        <div>
          <div
            class="SelectboxContainer"
          >
            <button
              class="buttonContainer  buttonIcon"
            >
              <span>
                Sıralama
              </span>
              <img
                alt="buttonIcon"
                src="DownArrow.svg"
              />
            </button>
            <div
              class="SelectboxOptions"
            />
          </div>
        </div>
      </div>
      <div
        class="filterAndProducts"
      >
        <div
          class="sideBarContainer"
        >
          <div
            class="SidebarContainer"
          >
            <p
              class="FilterName"
            >
              Renk
            </p>
            <ul
              class="FilterList"
            >
              <li
                class="Element "
              >
                <span>
                  Siyah
                </span>
                <span>
                  (
                  1
                  )
                </span>
              </li>
            </ul>
          </div>
          <div
            class="SidebarContainer"
          >
            <p
              class="SortName"
            >
              Sıralama
            </p>
            <ul
              class="SortList"
            />
          </div>
          <div
            class="SidebarContainer"
          >
            <p
              class="FilterName"
            >
              Marka
            </p>
            <ul
              class="FilterList"
            >
              <li
                class="Element "
              >
                <span>
                  Apple
                </span>
                <span>
                  (
                  1
                  )
                </span>
              </li>
            </ul>
          </div>
        </div>
        <div>
          <div
            class="productListContainer"
          >
            <div
              class="productCardContainer"
            >
              <div
                class="imageContainer"
              >
                <img
                  class="image"
                  src="https://productimages.hepsiburada.net/s/49/400-592/10986385899570.jpg"
                />
              </div>
              <div
                class="productProperties"
              >
                <div>
                  <p
                    class="productTitle"
                  >
                    iPhone 11 64GB
                  </p>
                  <p
                    class="productbrand"
                  >
                    <span>
                      Marka: 
                    </span>
                    Apple
                  </p>
                  <p
                    class="productColor"
                  >
                    <span>
                      Renk: 
                    </span>
                    Siyah
                  </p>
                  <p
                    class="discountedPrice"
                  >
                    12.276,00
                     
                    TL
                  </p>
                </div>
                <div
                  class="prevPriceContainer"
                >
                  <span
                    class="prevPrice"
                  >
                    13.950,00
                     TL
                  </span>
                  <span
                    class="discountRate"
                  >
                    12
                    %
                  </span>
                </div>
                <div
                  class="productCartButton"
                >
                  <button
                    class="buttonContainer orange undefined"
                    disabled=""
                  >
                    <span>
                      Bu ürün zaten sepetinizde ekli
                    </span>
                  </button>
                </div>
              </div>
            </div>
          </div>
          <div>
            <div
              class="PaginationContainer"
            >
              <span
                class="arrowButton FirstArrow"
              >
                &lt;&lt;
              </span>
              <span
                class="arrowButton"
              >
                &lt;
              </span>
              <ul
                class="PaginationList"
              >
                <li
                  class="activeNumber"
                >
                  <span
                    class="numberButton"
                  >
                    1
                  </span>
                </li>
              </ul>
              <span
                class="arrowButton"
              >
                &gt;
              </span>
              <span
                class="arrowButton LastArrow"
              >
                &gt;&gt;
              </span>
            </div>
          </div>
        </div>
      </div>
    </div>
  </div>,
  "debug": [Function],
  "findAllByAltText": [Function],
  "findAllByDisplayValue": [Function],
  "findAllByLabelText": [Function],
  "findAllByPlaceholderText": [Function],
  "findAllByRole": [Function],
  "findAllByTestId": [Function],
  "findAllByText": [Function],
  "findAllByTitle": [Function],
  "findByAltText": [Function],
  "findByDisplayValue": [Function],
  "findByLabelText": [Function],
  "findByPlaceholderText": [Function],
  "findByRole": [Function],
  "findByTestId": [Function],
  "findByText": [Function],
  "findByTitle": [Function],
  "getAllByAltText": [Function],
  "getAllByDisplayValue": [Function],
  "getAllByLabelText": [Function],
  "getAllByPlaceholderText": [Function],
  "getAllByRole": [Function],
  "getAllByTestId": [Function],
  "getAllByText": [Function],
  "getAllByTitle": [Function],
  "getByAltText": [Function],
  "getByDisplayValue": [Function],
  "getByLabelText": [Function],
  "getByPlaceholderText": [Function],
  "getByRole": [Function],
  "getByTestId": [Function],
  "getByText": [Function],
  "getByTitle": [Function],
  "queryAllByAltText": [Function],
  "queryAllByDisplayValue": [Function],
  "queryAllByLabelText": [Function],
  "queryAllByPlaceholderText": [Function],
  "queryAllByRole": [Function],
  "queryAllByTestId": [Function],
  "queryAllByText": [Function],
  "queryAllByTitle": [Function],
  "queryByAltText": [Function],
  "queryByDisplayValue": [Function],
  "queryByLabelText": [Function],
  "queryByPlaceholderText": [Function],
  "queryByRole": [Function],
  "queryByTestId": [Function],
  "queryByText": [Function],
  "queryByTitle": [Function],
  "rerender": [Function],
  "unmount": [Function],
}
`;
